From 43a283592a5d8513165ff11469921dd81520d75e Mon Sep 17 00:00:00 2001
From: Venkataramana Kotakonda <venkataramana.kotakonda@intel.com>
Date: Wed, 26 Aug 2015 21:16:08 +0530
Subject: [PATCH] USBC-DispPe: Send DI NAK during device mode.

In device mode, when host requests for discover identity, device
should send NAK as display streaming is not supported in device mode.

Change-Id: Ib939a0a2a7cb782ca871900b7d35869efeb09704
Tracked-On: https://jira01.devtools.intel.com/browse/GMINL-17330
Signed-off-by: Venkataramana Kotakonda <venkataramana.kotakonda@intel.com>
---
 drivers/usb/typec/pd/display_pe.c | 19 ++++++++++++++++++-
 1 file changed, 18 insertions(+), 1 deletion(-)

diff --git a/drivers/usb/typec/pd/display_pe.c b/drivers/usb/typec/pd/display_pe.c
index cfbd203cced7..18d349a49dee 100644
--- a/drivers/usb/typec/pd/display_pe.c
+++ b/drivers/usb/typec/pd/display_pe.c
@@ -174,6 +174,22 @@ static int disp_pe_send_discover_identity(struct disp_port_pe *disp_pe)
 	return ret;
 }
 
+static int disp_pe_send_discover_identity_rnak(struct disp_port_pe *disp_pe)
+{
+	int ret;
+	struct vdm_header v_hdr = { 0 };
+
+	v_hdr.cmd = DISCOVER_IDENTITY;
+	v_hdr.cmd_type = REP_NACK;
+	v_hdr.vdm_type = STRUCTURED_VDM; /* Structured VDM */
+	v_hdr.svid = PD_SID;
+
+	ret = policy_send_packet(&disp_pe->p, &v_hdr, 4,
+				PD_DATA_MSG_VENDOR_DEF, PE_EVT_SEND_VDM);
+
+	return ret;
+}
+
 static int disp_pe_send_discover_svid(struct disp_port_pe *disp_pe)
 {
 	struct pd_packet pkt;
@@ -243,7 +259,8 @@ static int disp_pe_handle_discover_identity(struct disp_port_pe *disp_pe,
 
 	switch (vdm_hdr->cmd_type) {
 	case INITIATOR:
-		log_warn("UFP alternate mode not supported\n");
+		log_warn("UFP alternate mode not supported, Sending NAK\n");
+		disp_pe_send_discover_identity_rnak(disp_pe);
 		break;
 	case REP_ACK:
 		if ((disp_pe->state != DISP_PE_STATE_DI_SENT)
-- 
1.9.1

